Add extended mnemonics

dominik-steenken opened the pull request:
This PR adds a number of thus-far missing extended mnemonics to the assembler and disassembler for SystemZ.

The following mnemonics have been added and are supported for the assembler and disassembler:

- `NOP(R)?`
- `LFI`
- `RISBG(N)?Z`

The following mnemonics have been added and are supported for the assembler only:

- `JC(TH)?`
- `LLG(F|H)I`
- `NOT(G)?R`

dominik-steenken committed:
[SystemZ] Support Disassembly of NOP and NOPR
This is accomplished by two additional instruction formats, representing
conditional branch instructions whose condition is never true.

dominik-steenken committed:
[SystemZ] Add missing instruction / mnemonic aliases
This adds aliases JC, JCTH, LFI, LLGFI and LLGHI

dominik-steenken committed:
[SystemZ] Add NOTR and NOTGR for the AsmParser only
This necessitates a new InstructionFormat (UnaryRRFa). The instructions are
not made available for the disassembler because distinguishing their encoding
from encodings of other instructions (NORK and NOGRK) requires a check on the
register choice (R2 == R3) that, to my knowledge, the disassembler is currently
not capable of.

dominik-steenken committed:
[SystemZ] Introduce handling of T and Z bits for the RIE-f instruction format
Also add two extended mnemonics that use this feature, RISBGZ and RISBGNZ.
T and Z bit handling are provided by extending the RIE-f format with two
optional `bits<8>` parameters that OR their content onto the I3 and I4 fields.
Setting these to `0x80` ensures that the highest bit of I3 or I4 will be set,
thus setting the T or Z bit, respectively.

redstar commented on llvm/lib/Target/SystemZ/SystemZInstrInfo.td:
```diff
@@ -973,6 +980,7 @@ let isAsCheapAsAMove = 1, isMoveImm = 1, isReMaterializable = 1 in {
   def IILF : UnaryRIL<"iilf", 0xC09, bitconvert, GR32, uimm32>;
   def IIHF : UnaryRIL<"iihf", 0xC08, bitconvert, GRH32, uimm32>;
 }
+def LFI : InstAlias<"lfi\t$R1, $RI1",(IILF GR32:$R1, uimm32:$RI1)>;
```
```suggestion
def LFI : InstAlias<"lfi\t$R1, $RI1",(IILF GR32:$R1, uimm32:$RI1)>;
```
```suggestion
def LFI : InstAlias<"lfi\t$R1, $RI1", (IILF GR32:$R1, uimm32:$RI1)>;
```

dominik-steenken replied:
fixed

redstar commented on llvm/lib/Target/SystemZ/SystemZInstrInfo.td:
```diff
@@ -464,6 +469,8 @@ let isAsCheapAsAMove = 1, isMoveImm = 1, isReMaterializable = 1 in {
   def LLILF : UnaryRIL<"llilf", 0xC0F, bitconvert, GR64, imm64lf32>;
   def LLIHF : UnaryRIL<"llihf", 0xC0E, bitconvert, GR64, imm64hf32>;
 }
+def LLGFI : InstAlias<"llgfi\t$R1, $RI1",(LLILF GR64:$R1, imm64lf32:$RI1)>;
+def LLGHI : InstAlias<"llghi\t$R1, $RI1",(LLILL GR64:$R1, imm64ll16:$RI1)>;
```
```suggestion
def LLGFI : InstAlias<"llgfi\t$R1, $RI1", (LLILF GR64:$R1, imm64lf32:$RI1)>;
def LLGHI : InstAlias<"llghi\t$R1, $RI1", (LLILL GR64:$R1, imm64ll16:$RI1)>;
```
```suggestion
def LLGFI : InstAlias<"llgfi\t$R1, $RI1",(LLILF GR64:$R1, imm64lf32:$RI1)>;
def LLGHI : InstAlias<"llghi\t$R1, $RI1",(LLILL GR64:$R1, imm64ll16:$RI1)>;
```

dominik-steenken replied:
fixed

redstar commented on llvm/lib/Target/SystemZ/SystemZInstrInfo.td:
```diff
@@ -111,11 +111,16 @@ let isBranch = 1, isTerminator = 1, isBarrier = 1 in {
-def NOPR_bare : InstAlias<"nopr", (BCRAsm 0, R0D), 0>;
+def NOPR : NeverCondBranchRR<"nopr", 0x07>;
+def NOPR_bare : InstAlias<"nopr", (NOPR R0D), 0>;
+// def NOP  : InstAlias<"nop\t$XBD", (BCAsm 0, bdxaddr12only:$XBD), 0>;
+// let isAsmParserOnly = 1, hasNoSchedulingInfo = 1, M1 = 0, X2 = 0, B2 = 0, D2 = 0 in
+//   def NOP_bare  : InstRXb<0x47,(outs), (ins), "nop", []>;
+// def NOPR : InstAlias<"nopr\t$R", (BCRAsm 0, GR64:$R), 0>;
+// def NOPR_bare : InstAlias<"nopr", (BCRAsm 0, R0D), 0>;
```
No dead code in comments please.

dominik-steenken replied:
goddamnit. removed.

redstar commented on llvm/lib/Target/SystemZ/SystemZInstrFormats.td:
```diff
@@ -363,7 +363,7 @@ class InstRIEe<bits<16> op, dag outs, dag ins, string asmstr, list<dag> pattern>
 }
 
-class InstRIEf<bits<16> op, dag outs, dag ins, string asmstr, list<dag> pattern>
+class InstRIEf<bits<16> op, dag outs, dag ins, string asmstr, list<dag> pattern, bits<8> I3Or = 0, bits<8> I4Or = 0>
```
Not sure how serious this is applied to .td files, but the line is way over the 80 char limit. (Well, the original line is a little bit too long, too.)

dominik-steenken replied:
I split the line, similar to how it is done in the handful of other formats with extra parms.

redstar reviewed (commented):
Small style issue.

dominik-steenken committed:
Clean up formatting and remove comments

uweigand commented:
The CI shows failures in insn-good.s and a bunch of SystemZ LLD tests, these definitely need to be addressed.

Also, can you re-check that all mnemonics added here are also accepted by GAS?  I think at least these may not be:  nhhr, nhlr, nlhr, ohhr, ohlr, olhr, xhhr, xhlr, xlhr.   You should verify that GAS can assemble all the insn-good* files (resulting in mostly identical object files - at least the disassembly of those object files needs to match again).

dominik-steenken committed:
Fix insn-good.s test

dominik-steenken committed:
Fix linker tests
These tests were expecting the old `bc 0, 0` / `bcr 0, %r0`
disassemblies for nops. This commit changes them to expect
the proper mnemonics in disassembly.

dominik-steenken commented:
> The CI shows failures in insn-good.s and a bunch of SystemZ LLD tests, these definitely need to be addressed.

I have fixed those tests. In `insn-good.s` some Aliases were pinned to the wrong target instruction, and the *lld* tests needed to be modified to account for `nop` and `nopr` now being emitted by the disassembler.

> Also, can you re-check that all mnemonics added here are also accepted by GAS? I think at least these may not be: nhhr, nhlr, nlhr, ohhr, ohlr, olhr, xhhr, xhlr, xlhr. You should verify that GAS can assemble all the insn-good* files (resulting in mostly identical object files - at least the disassembly of those object files needs to match again).

I moved the extended mnemonics of `R(N|O|X)SBG` to a different PR and re-checked that gas can assemble all of our assembly tests. While doingf that, we discovered an issue with register types in some instructions, as well as a difference in permissiveness regarding the immediate fields with `Z` and `T` bits that will be addressed in separate PRs.

uweigand reviewed (approved):
LGTM now, thanks!

dominik-steenken commented:
follow-up build to the above is clean
